### PR TITLE
[AArch64][GlobalISel] Add support for pre-indexed loads/stores.

### PR DESCRIPTION
**Pull request by aemerson**
The pre-index matcher just needs some small heuristics to make sure it doesn't cause regressions. Apart from that it's a simple change, since the only difference is an immediate operand of '1' vs '0' in the instruction.

Part of a stack https://github.com/llvm/llvm-project/pull/69533

### CODE REVIEW

**Comment by tschuett** [comment]
You added new asserts without the common && "something bad happened". I am wondering why they are not just 
if() return false?

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -1141,11 +1141,29 @@ bool CombinerHelper::findPreIndexCandidate(GLoadStore &LdSt, Register &Addr,
...
+      return false;
+  }
+
   // FIXME: check whether all uses of the base pointer are constant PtrAdds.
   // That might allow us to end base's liveness here by adjusting the constant.
-
-  return all_of(MRI.use_nodbg_instructions(Addr),
-                [&](MachineInstr &UseMI) { return dominates(LdSt, UseMI); });
+  bool RealUse = false;
+  for (auto &PtrUse : MRI.use_nodbg_instructions(Addr)) {
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:1153]: 10 lines up it is called AddrUse?!?

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 2]
```diff
@@ -1141,11 +1141,29 @@ bool CombinerHelper::findPreIndexCandidate(GLoadStore &LdSt, Register &Addr,
       return false;
   }
 
+  // Avoid increasing cross-block register pressure.
+  for (auto &AddrUse : MRI.use_nodbg_instructions(Addr)) {
+    if (AddrUse.getParent() != LdSt.getParent())
+      return false;
+  }
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:1148]: Are you sure about the braces?

**Comment by aemerson** [comment]
> You added new asserts without the common && "something bad happened". I am wondering why they are not just if() return false?

It's something we're not that consistent with. In selection of instructions where we should be confident in the completeness of the legalization rules, we can use asserts. However if the legalizer rules may have gaps, it's probably better to be safe and fall back instead of asserting.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -1141,11 +1141,28 @@ bool CombinerHelper::findPreIndexCandidate(GLoadStore &LdSt, Register &Addr,
...
+      return false;
+
   // FIXME: check whether all uses of the base pointer are constant PtrAdds.
   // That might allow us to end base's liveness here by adjusting the constant.
-
-  return all_of(MRI.use_nodbg_instructions(Addr),
-                [&](MachineInstr &UseMI) { return dominates(LdSt, UseMI); });
+  bool RealUse = false;
+  for (auto &AddrUse : MRI.use_nodbg_instructions(Addr)) {
+    if (!dominates(LdSt, AddrUse))
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:1153]: Defer the dominance check until after you know the mode is foldable?

File: llvm/lib/Target/AArch64/AArch64Combine.td [hunk 4]
```diff
@@ -245,7 +245,7 @@ def AArch64PostLegalizerLowering
 // Post-legalization combines which are primarily optimizations.
 def AArch64PostLegalizerCombiner
     : GICombiner<"AArch64PostLegalizerCombinerImpl",
-                       [copy_prop, combines_for_extload,
+                       [copy_prop, combines_for_extload, reassocs,
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/AArch64Combine.td:248]: Looks unrelated?

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 5]
```diff
@@ -5659,24 +5659,34 @@ bool AArch64InstructionSelector::selectIndexedLoad(MachineInstr &MI,
...
-  static constexpr unsigned GPROpcodes[] = {
-      AArch64::LDRBBpost, AArch64::LDRHHpost, AArch64::LDRWpost,
-      AArch64::LDRXpost};
-  static constexpr unsigned FPROpcodes[] = {
-      AArch64::LDRBpost, AArch64::LDRHpost, AArch64::LDRSpost,
-      AArch64::LDRDpost, AArch64::LDRQpost};
-
+  LLT Ty = MRI.getType(Dst);
+  assert(Ty.getSizeInBits() <= 128 && "Unexpected type for indexed load");
   unsigned MemSize = Ld.getMMO().getMemoryType().getSizeInBytes();
```
Comment (jroelofs) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5664]: dumb question: are we guaranteed that `Ty.getSizeInBits() == MemSize`, or could the size of the memory operand be wider than what's actually loaded?

Comment (aemerson) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:5664]: Good question. We have separate opcodes for extending loads, `G_INDEXED_ZEXTLOAD/G_INDEXED_SEXTLOAD`. If the load's destination type is > memory type here it's an any extending load, so we just use the memory type to decide which instruction to pick.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp (hunk 3 above)
Comment (aemerson) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:1153]: The foldable check isn't an early exit, so we still need to do this for every use.

File: llvm/lib/Target/AArch64/AArch64Combine.td (hunk 4 above)
Comment (aemerson) [reply llvm/lib/Target/AArch64/AArch64Combine.td:248]: Yep, will remove.

**Review by jroelofs — APPROVED** [review]
LGTM
